File: iot_driver_copilot/logitech_camera/driver.swift
import os
import cv2
import threading
import time
import tempfile
from flask import Flask, Response, jsonify, request, send_file

app = Flask(__name__)

CAMERA_LOCK = threading.Lock()
CAMERA_MANAGER_LOCK = threading.Lock()

DEFAULT_RESOLUTION = (
    int(os.environ.get("CAMERA_RES_WIDTH", 640)),
    int(os.environ.get("CAMERA_RES_HEIGHT", 480)),
)
DEFAULT_FPS = int(os.environ.get("CAMERA_FPS", 20))
SERVER_HOST = os.environ.get("SERVER_HOST", "0.0.0.0")
SERVER_PORT = int(os.environ.get("SERVER_PORT", 8080))

class CameraSession:
    def __init__(self, camera_id=0, resolution=DEFAULT_RESOLUTION, fps=DEFAULT_FPS):
        self.camera_id = camera_id
        self.resolution = resolution
        self.fps = fps
        self.cap = None
        self.active = False

    def start(self):
        with CAMERA_LOCK:
            if self.cap is not None:
                return
            self.cap = cv2.VideoCapture(self.camera_id)
            self.cap.set(cv2.CAP_PROP_FRAME_WIDTH, self.resolution[0])
            self.cap.set(cv2.CAP_PROP_FRAME_HEIGHT, self.resolution[1])
            self.cap.set(cv2.CAP_PROP_FPS, self.fps)
            if not self.cap.isOpened():
                self.cap = None
                raise RuntimeError("Failed to open camera {}".format(self.camera_id))
            self.active = True

    def stop(self):
        with CAMERA_LOCK:
            if self.cap is not None:
                self.cap.release()
                self.cap = None
                self.active = False

    def read(self):
        with CAMERA_LOCK:
            if self.cap is None or not self.cap.isOpened():
                raise RuntimeError("Camera not started")
            ret, frame = self.cap.read()
            if not ret:
                raise RuntimeError("Failed to capture frame")
            return frame

    def is_active(self):
        return self.cap is not None and self.cap.isOpened()

class CameraManager:
    def __init__(self):
        self.sessions = {}
        self.selected_id = None
        self.list_cameras()

    def list_cameras(self):
        cameras = []
        for i in range(10):
            cap = cv2.VideoCapture(i)
            if cap.isOpened():
                cameras.append(i)
                cap.release()
        return cameras

    def get_status(self):
        info = []
        for cam_id in self.list_cameras():
            session = self.sessions.get(cam_id)
            info.append({
                "camera_id": cam_id,
                "active": bool(session and session.is_active())
            })
        return info

    def start_camera(self, camera_id=None):
        with CAMERA_MANAGER_LOCK:
            if camera_id is None:
                camera_id = 0
            if camera_id not in self.list_cameras():
                raise RuntimeError("Camera ID {} not found".format(camera_id))
            if camera_id not in self.sessions:
                self.sessions[camera_id] = CameraSession(camera_id)
            self.sessions[camera_id].start()
            self.selected_id = camera_id

    def stop_camera(self):
        with CAMERA_MANAGER_LOCK:
            if self.selected_id is not None and self.selected_id in self.sessions:
                self.sessions[self.selected_id].stop()
                self.selected_id = None

    def select_camera(self, camera_id):
        with CAMERA_MANAGER_LOCK:
            if camera_id not in self.list_cameras():
                raise RuntimeError("Camera ID {} not found".format(camera_id))
            self.stop_camera()
            self.start_camera(camera_id)

    def get_active_session(self):
        if self.selected_id is not None and self.selected_id in self.sessions:
            return self.sessions[self.selected_id]
        return None

camera_manager = CameraManager()

def mjpeg_stream(session):
    try:
        while session.is_active():
            frame = session.read()
            ret, jpeg = cv2.imencode('.jpg', frame)
            if not ret:
                continue
            yield (b'--frame\r\n'
                   b'Content-Type: image/jpeg\r\n\r\n' + jpeg.tobytes() + b'\r\n')
            time.sleep(1.0 / session.fps)
    except Exception:
        pass

@app.route("/cameras", methods=["GET"])
def list_cameras():
    return jsonify({
        "cameras": camera_manager.get_status()
    })

@app.route("/camera/start", methods=["POST"])
def start_camera():
    try:
        args = request.json if request.is_json else {}
        camera_id = args.get("camera_id", 0)
        camera_manager.start_camera(camera_id)
        return jsonify({"status": "started", "camera_id": camera_manager.selected_id})
    except Exception as e:
        return jsonify({"error": str(e)}), 400

@app.route("/camera/stop", methods=["POST"])
def stop_camera():
    try:
        camera_manager.stop_camera()
        return jsonify({"status": "stopped"})
    except Exception as e:
        return jsonify({"error": str(e)}), 400

@app.route("/cameras/select", methods=["PUT"])
def select_camera():
    try:
        data = request.json
        if not data or "camera_id" not in data:
            return jsonify({"error": "Missing camera_id"}), 400
        camera_manager.select_camera(int(data["camera_id"]))
        return jsonify({"status": "switched", "camera_id": camera_manager.selected_id})
    except Exception as e:
        return jsonify({"error": str(e)}), 400

@app.route("/camera/stream", methods=["GET"])
def camera_stream():
    session = camera_manager.get_active_session()
    if not session or not session.is_active():
        return jsonify({"error": "No camera started"}), 400
    return Response(mjpeg_stream(session),
                    mimetype='multipart/x-mixed-replace; boundary=frame')

@app.route("/camera/capture", methods=["GET"])
def camera_capture():
    try:
        session = camera_manager.get_active_session()
        if not session or not session.is_active():
            return jsonify({"error": "No camera started"}), 400
        frame = session.read()
        ret, jpeg = cv2.imencode('.jpg', frame)
        if not ret:
            raise Exception("Failed to encode image")
        temp = tempfile.NamedTemporaryFile(delete=False, suffix=".jpg")
        temp.write(jpeg.tobytes())
        temp.close()
        meta = {
            "width": frame.shape[1],
            "height": frame.shape[0],
            "channels": frame.shape[2] if len(frame.shape) > 2 else 1,
            "camera_id": session.camera_id
        }
        resp = send_file(temp.name, mimetype="image/jpeg")
        resp.headers['X-Camera-Meta'] = str(meta)
        os.unlink(temp.name)
        return resp
    except Exception as e:
        return jsonify({"error": str(e)}), 400

@app.route("/camera/record", methods=["POST"])
def camera_record():
    try:
        session = camera_manager.get_active_session()
        if not session or not session.is_active():
            return jsonify({"error": "No camera started"}), 400
        data = request.json
        duration = int(data.get("duration", 5)) if data else 5
        ext = data.get("format", "mp4")
        fourcc = cv2.VideoWriter_fourcc(*("mp4v" if ext == "mp4" else "XVID"))
        temp = tempfile.NamedTemporaryFile(delete=False, suffix="." + ext)
        out = cv2.VideoWriter(temp.name, fourcc, session.fps, session.resolution)
        start_time = time.time()
        try:
            while time.time() - start_time < duration:
                with CAMERA_LOCK:
                    frame = session.read()
                    out.write(frame)
                time.sleep(1.0 / session.fps)
        finally:
            out.release()
        resp = send_file(temp.name, mimetype="video/mp4" if ext == "mp4" else "video/avi")
        os.unlink(temp.name)
        return resp
    except Exception as e:
        return jsonify({"error": str(e)}), 400

if __name__ == "__main__":
    app.run(host=SERVER_HOST, port=SERVER_PORT, threaded=True)

File: iot_driver_copilot/logitech_camera/driver.py
import os
import io
import cv2
import time
import threading
from datetime import datetime
from flask import Flask, Response, jsonify, request, send_file, stream_with_context

# Environment Variables
HTTP_HOST = os.environ.get("HTTP_HOST", "0.0.0.0")
HTTP_PORT = int(os.environ.get("HTTP_PORT", "8080"))
CAMERA_INDEX = int(os.environ.get("CAMERA_INDEX", "0"))  # USB camera index (default 0)
FRAME_WIDTH = int(os.environ.get("FRAME_WIDTH", "640"))
FRAME_HEIGHT = int(os.environ.get("FRAME_HEIGHT", "480"))
IMAGE_FORMAT = os.environ.get("IMAGE_FORMAT", "jpeg").lower()  # jpeg or png

app = Flask(__name__)

# Globals for streaming control
streaming = False
streaming_lock = threading.Lock()
camera = None

def initialize_camera():
    global camera
    if camera is None:
        cam = cv2.VideoCapture(CAMERA_INDEX)
        cam.set(cv2.CAP_PROP_FRAME_WIDTH, FRAME_WIDTH)
        cam.set(cv2.CAP_PROP_FRAME_HEIGHT, FRAME_HEIGHT)
        if not cam.isOpened():
            raise RuntimeError("Unable to open camera at index %d" % CAMERA_INDEX)
        camera = cam
    return camera

def release_camera():
    global camera
    if camera is not None:
        camera.release()
        camera = None

def get_image(format='jpeg'):
    cam = initialize_camera()
    ret, frame = cam.read()
    if not ret:
        raise RuntimeError("Failed to capture image from camera")
    img_ext = '.jpg' if format == 'jpeg' else '.png'
    encode_param = [int(cv2.IMWRITE_JPEG_QUALITY), 95] if format == 'jpeg' else [int(cv2.IMWRITE_PNG_COMPRESSION), 3]
    ret2, img = cv2.imencode(img_ext, frame, encode_param)
    if not ret2:
        raise RuntimeError("Failed to encode image")
    return img.tobytes(), format

def mjpeg_stream_gen():
    global streaming
    cam = initialize_camera()
    try:
        while True:
            with streaming_lock:
                if not streaming:
                    break
            ret, frame = cam.read()
            if not ret:
                continue
            ret2, jpeg = cv2.imencode('.jpg', frame, [int(cv2.IMWRITE_JPEG_QUALITY), 80])
            if not ret2:
                continue
            yield (b'--frame\r\n'
                   b'Content-Type: image/jpeg\r\n\r\n' + jpeg.tobytes() + b'\r\n')
    finally:
        pass  # Do not release camera here, may be reused

@app.route('/camera/info', methods=['GET'])
def camera_info():
    return jsonify({
        "device_name": "Logitech Camera",
        "device_model": "Logitech Camera",
        "manufacturer": "Logitech",
        "device_type": "Camera",
        "supported_formats": ["MJPEG", "YUYV", "H.264"],
        "image_formats": ["JPEG", "PNG"],
        "streaming_endpoint": "/stream/start (POST), /camera/stream/start (POST)",
        "capture_endpoint": "/capture (POST), /camera/capture (POST)"
    })

@app.route('/camera/capture', methods=['POST'])
@app.route('/capture', methods=['POST'])
def capture_image():
    try:
        fmt = request.args.get("format", IMAGE_FORMAT)
        img_bytes, fmt = get_image(fmt)
        now = datetime.utcnow().isoformat() + "Z"
        filename = f"capture_{int(time.time())}.{fmt}"
        buf = io.BytesIO(img_bytes)
        buf.seek(0)
        return send_file(
            buf,
            mimetype=f"image/{fmt}",
            as_attachment=True,
            download_name=filename,
            etag=now,
            last_modified=time.time(),
            add_etags=True,
            conditional=True
        ), 200, {
            "X-Image-Format": fmt,
            "X-Timestamp": now
        }
    except Exception as e:
        return jsonify({"error": str(e)}), 500

@app.route('/stream/start', methods=['POST'])
@app.route('/camera/stream/start', methods=['POST'])
def start_stream():
    global streaming
    with streaming_lock:
        if streaming:
            return jsonify({"status": "already streaming", "url": "/stream/video"}), 200
        streaming = True
    # Do not start the actual stream in background; the client will GET /stream/video
    return jsonify({"status": "streaming started", "stream_url": "/stream/video"}), 200

@app.route('/stream/stop', methods=['POST'])
@app.route('/camera/stream/stop', methods=['POST'])
def stop_stream():
    global streaming
    with streaming_lock:
        if not streaming:
            return jsonify({"status": "not streaming"}), 200
        streaming = False
    time.sleep(0.2)
    return jsonify({"status": "streaming stopped"}), 200

@app.route('/stream/video', methods=['GET'])
def stream_video():
    global streaming
    with streaming_lock:
        if not streaming:
            return jsonify({"error": "stream not started. POST /stream/start first."}), 400
    return Response(stream_with_context(mjpeg_stream_gen()),
                    mimetype='multipart/x-mixed-replace; boundary=frame')

@app.route('/', methods=['GET'])
def root():
    return jsonify({
        "message": "Logitech Camera Driver",
        "endpoints": [
            {"path": "/camera/info", "method": "GET", "description": "Camera information"},
            {"path": "/camera/capture", "method": "POST", "description": "Capture image"},
            {"path": "/capture", "method": "POST", "description": "Capture image"},
            {"path": "/stream/start", "method": "POST", "description": "Start streaming"},
            {"path": "/camera/stream/start", "method": "POST", "description": "Start streaming"},
            {"path": "/stream/stop", "method": "POST", "description": "Stop streaming"},
            {"path": "/camera/stream/stop", "method": "POST", "description": "Stop streaming"},
            {"path": "/stream/video", "method": "GET", "description": "MJPEG video stream"}
        ]
    })

@app.teardown_appcontext
def cleanup(exception=None):
    release_camera()

if __name__ == '__main__':
    app.run(host=HTTP_HOST, port=HTTP_PORT, threaded=True)